### ISSUE TITLE
Question on implementing basic OCPP implementation

### ISSUE THREAD

**Issue by jbaasanjav9** [mobilityhouse/ocpp#90]
Hello,

I am trying to implement the basic ocpp2.0 implementation. 
When I try to implement "B05 - Set Variables" under the 2.2 Configuring charging station section, I had a problem with handling a request on the charging station side.

My code for the server-side is:

``` python
class ChargePoint(cp):
     ...
     async def set_variables(self, attribute_value, component_type, variable_type ):
     variable_data_list = [{'attributeValue': attribute_value, 'component': {'name':component_type}, 'variable': {'name':variable_type}}]
    set_variables_request = call.SetVariablesPayload(set_variable_data = variable_data_list)
    response = await self.call(set_variables_request)
    print(response)
```

My code for the client-side is:

``` python
class ChargePoint(cp):
   ...
   @on('ChangeConfiguration')
   def on_set_variables(self):
      return call_result.SetVariablesPayload(set_variable_result=[['Accepted', {"name": 'evse'}, {"name": 'lcd'}]]) 
```

Just for testing purposes, I hardcoded some values for the call_result.SetVariablesPayload.

When try running the program I get an error on the charge point side saying 

raise NotImplementedError(f"No handler for '{msg.action}' "
ocpp.exceptions.NotImplementedError: No handler for 'SetVariables' registered."</code>
   
I could not find a SetVariables action under the enums.py. That is why I put @on('ChangeConfiguration').  

So:
1) In general, how should we handle the actions or use cases that are not required part of a Call message? or actions or use cases that do not appear in enums.py?

2) If it is possible could you direct me on, what I should do to figure out the error that I am getting?

Thank you very much for your time.

**Comment by OrangeTux** [comment]
Hello @jbaasanjav9,

It seems that you are mixing an OCPP 2.0 server with a OCPP 1.6 client. `SetVariables` is an OCPP 2.0 message, while `ChangeConfiguration` is a OCPP 1.6 message. The versions are not compatible with each other.

You should change the client code to:

``` python
  @on('SetVariables')
   def on_set_variables(self):
      return call_result.SetVariablesPayload(set_variable_result=[['Accepted', {"name": 'evse'}, {"name": 'lcd'}]]) 
```

**Comment by jbaasanjav9** [comment]
Hello @OrangeTux,

Ohh I see, I was using Ctrl+Shift+F to find the keywords in enums.py file. Then I did not realize that it was directing me to the enums.py under ocpp/v16 folder. My bad.

The package does not provide enums.py file under ocpp/v20 folder. 

1) Does that mean we should create our own enums.py for v20?

1.2) If yes, can I just follow what is in the ocpp/v16/enums.py and make some changes in the keywords and put it under ocpp/v20 folder and work with it?

1.3) If no, how?

Thank you very much.

**Comment by OrangeTux** [comment]
1) If you want to use enums: yes. Note that the enums map to strings. So using the string instead of an enum would do the job as well. 

1.2) Ideally I would to add either all enums for OCPP 2.0 or none of them. Adding only a couple might be even more confusing for users.

1.3) All enum definitions are defined in the OCPP 2.0 JSON schemas. It shouldn't be to hard to write a script that iterates over the schemas and create the enums.
But as said, you don't necessarily need enums. You can use strings instead. The following code (for OCPP 1.6) is equal in function:

``` python
@on(Action.Heartbeat)
def on_heartbeat(self, *args, **kwargs):
    pass
``` 

```
@on("Heartbeat")
def on_heartbeat(self, *args, **kwargs):
    pass
```

**Comment by jbaasanjav9** [comment]
Hello @OrangeTux,

Thank you for the quick reply. Interestingly, '@on('SetVariables')' worked.  

Now, I think I have to understand how incoming call requests are identified by their respective call result replies.

For example: 
On server side 
<pre><code>class ChargePoint(cp):
     ...
     async def set_variables(self, attribute_value, component_type, variable_type ):
     variable_data_list = [{'attributeValue': attribute_value, 'component': {'name':component_type}, 'variable': {'name':variable_type}}]
     set_variables_request = call.SetVariablesPayload(set_variable_data = variable_data_list)
     response = await self.call(set_variables_request)
     print(response)
</code></pre>

On client side :
<pre><code>class ChargePoint(cp):
   ...
   @on('SetVariables')
   def on_set_variables(self):
      return call_result.SetVariablesPayload(set_variable_result=[['Accepted', {"name": 'evse'}, {"name": 'lcd'}]]) 
</code></pre>

1) So from the server-side, it just constructs a message wraps it up, validates, and send it right?

2) Then on the client-side, client receives it and how does the client know that "Ok this message is respect to SetVariables action"?
2.1) Moreover, how this 'SetVariables' string as a parameter to @on(' ') catches or identifies the message and knows that "Ok I should respond to this request"?
2.2) Or where does 'SetVariables' string is defined as it is a valid action string? 

3) If it is possible, could you elaborate more on this background mechanism?

I am sorry for being too wordy, but I guess it would help you to understand clearly.

Thanks a lot. 🙇

**Comment by OrangeTux** [comment]
1) Almost. You seem to imply that only the server can initiate a a call (which is analogous to a request). That is not true. OCPP is a bidirectional protocol. This means that both ends of the connection, the server and client, can initiate calls. For example, a BootNotification request must be send from the chargepoint (the client) to the server.

2) I'm not sure if I totally understand you questions here.  
As stated by [section 4.2.1 of the OCPP-J specification](https://github.com/mobilityhouse/ocpp/blob/master/docs/v16/ocpp-j-1.6-specification.pdf) a Call contains a field 'Action'. The value of of this field is a string defining the action that must be executed by the receiver. 
If you instantiate a `ChargePoint` the [constructor of `ChargePoint`](https://github.com/mobilityhouse/ocpp/blob/master/ocpp/charge_point.py#L104-L107) iterates over all methods decorated with `@on()` and `@after()` [to create a route map](https://github.com/mobilityhouse/ocpp/blob/6098908e6f7e51a355bcbeb3eda7479f73d6781c/ocpp/routing.py#L62-L114). 
In your example the server sends a Call with the Action field set to `SetVariables`. The ocpp library just filters the action out of the request and [checks the route map](https://github.com/mobilityhouse/ocpp/blob/master/ocpp/charge_point.py#L160-L164) if a handler has been registered for that action. This is implemented in [ocpp.ChargePoint._handle_call()](https://github.com/mobilityhouse/ocpp/blob/master/ocpp/charge_point.py#L149-L221). 
Of a handler has been found it is executed. The response is returned to the caller.

3) Did my answer on #2 also answered this question? If not, I'm happy to explain it in more detail.

**Comment by jbaasanjav9** [comment]
AHA 💡 
<pre><code>         call = Call(
            unique_id=str(self._unique_id_generator()),
            action=payload.__class__.__name__[:-7],
            payload=remove_nones(camel_case_payload)
        )</code></pre>

This section is exactly doing what you are talking about. It truncates the last 7 characters from the data class name which truncates "Payload" word then attach it to the action. 

Thank you very much. 🙏
